### PR TITLE
feat(contracts): upgrade validator system in `Colosseum`

### PR DESCRIPTION
**Pull request by seolaoh**
# Description

Implemented validator system upgrade from using `ValidatorPool` to using `ValidatorManager` in `Colosseum`.

Since some bugs are found in `ValidatorManager` and `AssetManager`, fixed them appropriately. I recommend to review this PR commit by commit because it includes some refactoring.

### CODE REVIEW

**Comment by Pangssu** [comment]
Add a validation process for `_minSlashingAmount` to constructor.

**Review by Pangssu — CHANGES_REQUESTED** [review]

File: packages/contracts/contracts/L1/Colosseum.sol [hunk 1]
```diff
@@ -785,25 +854,20 @@ contract Colosseum is Initializable, ISemver {
...
         // parentBeaconRoot is non-zero for Cancun block
         bytes32 blockHash = _publicInput.parentBeaconRoot != bytes32(0)
             ? Hashing.hashBlockHeaderCancun(_publicInput, _rlps)
             : Hashing.hashBlockHeaderShanghai(_publicInput, _rlps);
 
-        require(
-            _srcOutputRootProof.nextBlockHash == blockHash,
-            "Colosseum: the block hash from public input must be matched"
-        );
+        if (_srcOutputRootProof.nextBlockHash != blockHash) revert BlockHashMismatched();
```
Comment (0xHansLee) [inline packages/contracts/contracts/L1/Colosseum.sol:864]: This will give the same error with L838, is it ok?

**Comment by seolaoh** [comment]
> Add a validation process for `_minSlashingAmount` to constructor.

What do you mean for a validator process? To check what?

File: packages/contracts/contracts/L1/interfaces/IValidatorManager.sol [hunk 2]
```diff
@@ -21,8 +21,7 @@ interface IValidatorManager {
      * | Status            | initiated | started | MIN_REGISTER_AMOUNT | MIN_START_AMOUNT |
      * +-------------------+-----------+---------+---------------------+------------------+
      * | NONE              | X         | X       | X                   | X                |
-     * | INACTIVE          | O         | X       | X                   | O/X              |
-     * | IN_JAIL           | O         | O/X     | O/X                 | O/X              |
+     * | INACTIVE          | O         | O/X     | X                   | O/X              |
      * | ACTIVE            | O         | X       | O                   | X                |
      * | CAN_START         | O         | X       | O                   | O                |
      * | STARTED           | O         | O       | O                   | X                |
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/L1/interfaces/IValidatorManager.sol:27]: `ACTIVE` looks good to me, but `ELIGIBLE` looks quite ambiguous to me. So how about like below?

- `INACTIVE` -> `UNSTAKED`
- `ACTIVE` -> `REGISTERED`
- `CAN_START` -> `READY`
- `STARTED` -> `INACTIVE`
- `CAN_SUBMIT_OUTPUT` -> `ACTIVE`

File: packages/contracts/contracts/L1/Colosseum.sol [hunk 3]
```diff
@@ -308,44 +399,30 @@ contract Colosseum is Initializable, ISemver {
         bytes32 _l1BlockHash,
         uint256 _l1BlockNumber,
         bytes32[] calldata _segments
-    ) external outputNotFinalized(_outputIndex) {
-        require(_outputIndex > 0, "Colosseum: challenge for genesis output is not allowed");
+    ) external {
```
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/L1/Colosseum.sol:402]: Yes, I think it's duplicate check.

File: packages/contracts/contracts/L1/interfaces/IValidatorManager.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (0xHansLee) [reply packages/contracts/contracts/L1/interfaces/IValidatorManager.sol:27]: Or `NOT_REGISTERED`?

**Comment by Pangssu** [comment]
> > Add a validation process for `_minSlashingAmount` to constructor.
> 
> What do you mean for a validation process? To check what?

Ignore it. I realised that it can't check it in the AssetManager contract. `_minSlashingAmount <= ValidatorManager.MIN_START_AMOUNT`

File: packages/contracts/contracts/L1/Colosseum.sol [hunk 4]
```diff
@@ -577,20 +671,21 @@ contract Colosseum is Initializable, ISemver {
...
 
         // Rollback output root.
         L2_ORACLE.replaceL2Output(_outputIndex, _outputRoot, _asserter);
 
+        // Switch validator system after validator pool contract terminated.
+        if (L2_ORACLE.VALIDATOR_POOL().isTerminated(_outputIndex)) {
+            // Unjail asserter.
+            L2_ORACLE.VALIDATOR_MANAGER().tryUnjail(_asserter, true);
+        }
+
```
Comment (sm-stack) [inline packages/contracts/contracts/L1/Colosseum.sol:688]: Just a question, why is this part added?

File: packages/contracts/contracts/L1/interfaces/IValidatorManager.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/L1/interfaces/IValidatorManager.sol:27]: Do you mean it for `UNSTAKED`?

File: packages/contracts/contracts/L1/Colosseum.sol (hunk 4 above)
Comment (seolaoh) [reply packages/contracts/contracts/L1/Colosseum.sol:688]: It's because when malicious challenger has done `proveFault` and asserter has been jailed, security council should unjail the asserter when `dismissChallenge`.

File: packages/contracts/contracts/L1/interfaces/IValidatorManager.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/L1/interfaces/IValidatorManager.sol:27]: cc. @sm-stack validator client also needs to apply this change

[the remark this answers is not in the document]
Comment (seolaoh) [reply packages/contracts/contracts/L1/interfaces/IValidatorManager.sol:27]: I think `NOT_REGISTERED` means that the validator should re-register, but the validator can only delegate more than `MIN_REGISTER_AMOUNT`. So how about `EXITED`? Because if the validator undelegates the asset left less than `MIN_REGISTER_AMOUNT`, it means the validator is not willing to act as a validator.

[the remark this answers is not in the document]
Comment (0xHansLee) [reply packages/contracts/contracts/L1/interfaces/IValidatorManager.sol:27]: It looks good to me.

**Review by sm-stack — APPROVED** [review]
LGTM

**Review by Pangssu — APPROVED** [review]
